File: packages/php/src/Http/Requester.php
<?php

declare(strict_types=1);

namespace AILabTools\Http;

use AILabTools\Exception\ApiException;
use AILabTools\Response\BaseResponse;
use AILabTools\Response\ErrorDetail;
use GuzzleHttp\ClientInterface;
use JsonException;

final class Requester
{
    public function __construct(
        private readonly ClientInterface $client,
        private readonly string $apiKey,
        private readonly string $baseUrl,
        private readonly float $timeout,
    ) {
    }

    /**
     * @template T of BaseResponse
     * @param array<string, mixed> $params
     * @param callable(array<string, mixed>): T $responseFactory
     * @return T
     */
    public function request(
        string $method,
        string $path,
        array $params,
        bool $multipart,
        callable $responseFactory,
    ): BaseResponse {
        $url = rtrim($this->baseUrl, '/') . '/' . ltrim($path, '/');
        $options = [
            'headers' => ['ailabapi-api-key' => $this->apiKey],
            'http_errors' => false,
            'timeout' => $this->timeout,
        ];
        $resources = [];

        if (strtoupper($method) === 'GET') {
            $options['query'] = self::withoutNulls($params);
        } elseif ($multipart) {
            $parts = [];
            foreach (self::withoutNulls($params) as $name => $value) {
                if ($value instanceof FileInput) {
                    $upload = $value->toMultipartPart($name);
                    $parts[] = $upload['part'];
                    if (is_resource($upload['close'])) {
                        $resources[] = $upload['close'];
                    }
                } else {
                    $parts[] = ['name' => $name, 'contents' => self::stringify($value)];
                }
            }
            $options['multipart'] = $parts;
        } else {
            $options['json'] = self::withoutNulls($params);
        }

        try {
            $response = $this->client->request($method, $url, $options);
            $statusCode = $response->getStatusCode();
            $raw = (string) $response->getBody();
        } finally {
            foreach ($resources as $resource) {
                if (is_resource($resource)) {
                    fclose($resource);
                }
            }
        }

        try {
            $payload = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
        } catch (JsonException) {
            throw new ApiException(
                $raw === '' ? 'Empty response' : 'Invalid JSON response',
                statusCode: $statusCode,
                raw: $raw,
            );
        }
        if (!is_array($payload)) {
            throw new ApiException('Invalid JSON response', statusCode: $statusCode, raw: $raw);
        }

        $errorCode = isset($payload['error_code']) ? (int) $payload['error_code'] : 0;
        if ($statusCode !== 200 || $errorCode !== 0) {
            $detail = isset($payload['error_detail']) && is_array($payload['error_detail'])
                ? ErrorDetail::fromArray($payload['error_detail'])
                : null;
            throw new ApiException(
                $detail->message ?? (isset($payload['error_msg']) ? (string) $payload['error_msg'] : 'AILabTools API Error'),
                statusCode: $statusCode,
                requestId: isset($payload['request_id']) ? (string) $payload['request_id'] : null,
                logId: isset($payload['log_id']) ? (string) $payload['log_id'] : null,
                detail: $detail,
                raw: $payload,
            );
        }

        return $responseFactory($payload);
    }

    /** @param array<string, mixed> $values
     *  @return array<string, mixed>
     */
    private static function withoutNulls(array $values): array
    {
        return array_filter($values, static fn (mixed $value): bool => $value !== null);
    }

    private static function stringify(mixed $value): string
    {
        if (is_bool($value)) {
            return $value ? 'true' : 'false';
        }
        if (is_array($value)) {
            return implode(',', array_map('strval', $value));
        }

        return (string) $value;
    }
}

File: packages/kotlin/src/main/kotlin/com/ailabtools/sdk/Requester.kt
package com.ailabtools.sdk

import java.io.IOException
import java.util.Collections
import java.util.concurrent.TimeUnit
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.DeserializationStrategy
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

internal class Requester(
    client: OkHttpClient,
    private val json: Json,
    private val apiKey: String,
    baseUrl: String,
    timeoutMillis: Long,
) {
    private val baseUrl = baseUrl.trimEnd('/')
    private val client = client.newBuilder()
        .connectTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
        .readTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
        .writeTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
        .callTimeout(timeoutMillis, TimeUnit.MILLISECONDS)
        .build()
    private val activeCalls = Collections.synchronizedSet(mutableSetOf<Call>())

    suspend fun <T> request(
        method: String,
        path: String,
        parameters: RequestParams,
        serializer: DeserializationStrategy<T>,
    ): T {
        val query = parameters.queryFields().filterValues { it != null }
        val body = parameters.bodyFields().filterValues { it != null }
        val urlBuilder = (baseUrl + "/" + path.trimStart('/')).toHttpUrl().newBuilder()
        query.forEach { (name, value) -> urlBuilder.addQueryParameter(name, stringify(value)) }

        val requestBuilder = Request.Builder()
            .url(urlBuilder.build())
            .header("ailabapi-api-key", apiKey)
        if (method.equals("GET", ignoreCase = true)) {
            requestBuilder.get()
        } else {
            val multipart = MultipartBody.Builder().setType(MultipartBody.FORM)
            body.forEach { (name, value) ->
                when (value) {
                    is FileInput -> multipart.addFormDataPart(
                        name,
                        value.filename,
                        value.bytes.toRequestBody(value.contentType.toMediaType()),
                    )
                    else -> multipart.addFormDataPart(name, stringify(value))
                }
            }
            requestBuilder.method(method, multipart.build())
        }

        val call = client.newCall(requestBuilder.build())
        activeCalls += call
        val response = try {
            call.await()
        } catch (error: IOException) {
            val reason = error.message ?: error.javaClass.simpleName
            throw AILabToolsException("AILabTools request failed: $reason", error)
        } finally {
            activeCalls -= call
        }

        response.use {
            val text = it.body?.string().orEmpty()
            val payload = try {
                json.parseToJsonElement(text).jsonObject
            } catch (error: Exception) {
                throw ApiException(
                    message = if (text.isBlank()) "Empty response" else "Invalid JSON response",
                    statusCode = it.code,
                )
            }
            validateResponse(it, payload)
            return try {
                json.decodeFromString(serializer, text)
            } catch (error: Exception) {
                throw AILabToolsException("Unable to deserialize AILabTools response", error)
            }
        }
    }

    fun close() {
        activeCalls.toList().forEach(Call::cancel)
    }

    private fun validateResponse(response: Response, payload: JsonObject) {
        val errorCode = payload["error_code"]?.jsonPrimitive?.intOrNull ?: 0
        if (response.isSuccessful && errorCode == 0) return
        val detail = payload["error_detail"]?.let {
            runCatching { json.decodeFromJsonElement(ErrorDetail.serializer(), it) }.getOrNull()
        }
        val message = detail?.message
            ?: payload["error_msg"]?.jsonPrimitive?.contentOrNull
            ?: "AILabTools API Error"
        throw ApiException(
            message = message,
            statusCode = response.code,
            requestId = payload["request_id"]?.jsonPrimitive?.contentOrNull,
            logId = payload["log_id"]?.jsonPrimitive?.contentOrNull,
            errorCode = errorCode,
            detail = detail,
            raw = payload,
        )
    }

    private fun stringify(value: Any?): String = when (value) {
        is Iterable<*> -> value.joinToString(",")
        is Array<*> -> value.joinToString(",")
        else -> value.toString()
    }

    private suspend fun Call.await(): Response = suspendCancellableCoroutine { continuation ->
        continuation.invokeOnCancellation { cancel() }
        enqueue(object : Callback {
            override fun onFailure(call: Call, error: IOException) {
                if (continuation.isActive) continuation.resumeWithException(error)
            }

            override fun onResponse(call: Call, response: Response) {
                if (continuation.isActive) {
                    continuation.resume(response)
                } else {
                    response.close()
                }
            }
        })
    }
}
